### ISSUE TITLE
flux-proxy leaves shell is bad shape after instance shutsdown

### ISSUE THREAD

**Issue by SteVwonder** [flux-framework/flux-core#2139]
If I start a flux instance, connect into it via `flux-proxy` (locally or via ssh), and then kill the flux instance, the shell that ran `flux-proxy` is left in bad shape.  I've had this happen under both bash and zsh.  Under bash, I can usually run `reset` and everything is ok.  Under zsh, the shell will happily consume characters and then display seemingly random characters as well as mutate previously typed characters. It is quite bizarre. [Video of the experience](https://gfycat.com/CreamyBriefDaddylonglegs).  This makes typing multi-character commands like `reset` or `exit` impossible.  I'm not sure exactly what is causing it, but I'm wondering if there is something we can do on our end to prevent zsh/bash from ending up in such a tragic state.

**Comment by grondo** [comment]
This was annoying me lately so I looked into it for a bit.

I don't think it is the shell from which you ran `flux-proxy` that ends up in bad shape, it is the current shell *spawned* by `flux-proxy` that becomes unusable.

This is actually a generic issue with (at least) bash when its parent is killed. E.g.:
```
 grondo@fluke108:~$ bash
 grondo@fluke108:~$ bash
 grondo@fluke108:~$ kill -9 $PPID
 grondo@fluke108:~$ Killed
 grondo@fluke108:~$ 
 grondo@fluke108:~$ llogout
Shared connection to fluke closed.
```

I'm guessing this is because when the parent of the subshell forcibly exits it is reparented to init and results in terminal confusion if not loss 

E.g., after `bash; bash` I see:
```
USER      PPID   PID  PGID  SESS CMD
grondo   13414 24790 24790 24790 -bash
grondo   24790 24906 24906 24790 bash
grondo   24906 24910 24910 24790 bash
```

If I kill my parent from the current foreground shell, the interactive shell is reparented to init:
```
grondo   13414 24790 24790 24790 -bash
grondo       1 24910 24910 24790 bash
```

And if I try to strace that shell (24910), it immediately gets `EIO` reading stdin, prints `exit` and exits:

```
 grondo@fluke108:/proc$ strace -p 24910
strace: Process 24910 attached
read(0, 0x7fffffffc9e7, 1)              = -1 EIO (Input/output error)
rt_sigprocmask(SIG_BLOCK, [INT], [], 8) = 0
[snip]
write(2, "exit\n", 5)                   = 5
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
setpgid(0, 24910)                       = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

I don't really understand why sometimes I am still connected to my *original* bash login shell, while other times the `exit` seems to log me completely out (??). In the first case, the login session is recoverable by `stty sane` or `reset` (sometimes)

For the case of `flux proxy`, the `flux-proxy` command is the shell's parent. `strace` after killing my flux session shows that the command is getting `EPIPE` reading or writing to now the terminated `rsh` processes, and is itself exiting. The shell then behaves as shown above.

From `strace`:
```
write(3, "\22\0\356\377\0\0\0001\0\tattr.get\0\20{\"name\":\"ran"..., 57) = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=25590, si_uid=6885} ---
write(3, "\22\0\356\377\0\0\0w\0\vlog.append\0T<24>1 2019"..., 127) = -1 EPIPE (Broken pipe)
--- SIGPIPE {si_signo=SIGPIPE, si_code=SI_USER, si_pid=25590, si_uid=6885} ---
close(6)                                = 0
close(3)                                = 0
shutdown(3, SHUT_WR)                    = -1 EBADF (Bad file descriptor)
wait4(25604, 0x7fffffffba3c, 0, NULL)   = -1 ECHILD (No child processes)
close(3)                                = -1 EBADF (Bad file descriptor)
munmap(0x2aaaaeba6000, 2365176)         = 0
close(8)                                = 0
close(7)                                = 0
unlink("/var/tmp/grondo/flux-proxy-Qy2cj7/local") = 0
rmdir("/var/tmp/grondo/flux-proxy-Qy2cj7") = 0
exit_group(0)                           = ?
+++ exited with 0 +++
```

The only thing we can probably do here is to have `flux-proxy` *not* exit on error writing to / reading from the proxy connection. Instead, maybe the command can emit an error (if that is possible) and allow the user to gracefully exit the subshell.

**Comment by grondo** [comment]
Another idea would be to have `flux proxy` work more like `ssh-agent`, i.e. with no args, the proxy FLUX_URI and perhaps flux-proxy pid are output to stdout:

```console
$ flux proxy ssh://hostname/tmp/flux-xxzzyy
FLUX_URI=local:///tmp/flux-proxy-zzz; export FLUX_URI;
FLUX_PROXY_PID=12345; export FLUX_PROXY_PID;
echo Proxy pid 12345;
$
```

This allows connecting the *current* shell to proxy with an `eval`
```
$ eval `flux proxy ssh://hostname/tmp/flux-xxzzyy`
Proxy pid 12345
$
```

However, with an argument, `flux-proxy` would follow the current behavior of exporting the correct environment and starting a child process (disconnecting when the child exits)

**Comment by garlick** [comment]
The comparison with `ssh-agent` is apt.  I wonder if we should try to make it work more like that overall, for principal of least surprise?

It might be nice to break out the "ssh connector helper" function into its own subcommand so the user-facing part can be worked on without breaking it.

**Comment by grondo** [comment]
One other issue with `flux proxy` is that it breaks the flux model of ensuring that the installation of flux being used is consistent with the instance to which you are connecting. That is, you can `flux proxy` with the flux installed on your desktop into a flux instance on a system running a completely different version of flux. A related issue is described in #1816.

Since `flux proxy` requires remote shell access anyway, what if instead of a tunnel for Flux local connector traffic, `flux proxy` (or perhaps a new command in case flux-proxy is useful in the future)  just logged the user into the remote system with the correct environment. We might have to add an attribute to the broker to specify the path to its `flux(1)` command, but then it could be as simple as running the user's shell under `flux env`.

**Comment by grondo** [comment]
I was hitting this again lately. I think it probably isn't the right approach to change how `flux proxy` works anymore.

However, I experimented with saving the terminal state and restoring it on error only, and it _seems_ to help.
I wonder if there is any potential harm to adding this?

```diff
diff --git a/src/cmd/builtin/proxy.c b/src/cmd/builtin/proxy.c
index 35a3ce7b8..ad5cc0c95 100644
--- a/src/cmd/builtin/proxy.c
+++ b/src/cmd/builtin/proxy.c
@@ -27,6 +27,7 @@
 #include <argz.h>
 #include <glob.h>
 #include <inttypes.h>
+#include <termios.h>
 
 #include "src/common/libutil/cleanup.h"
 #include "src/common/libutil/uri.h"
@@ -45,6 +46,36 @@ struct proxy_command {
 
 static const char *route_auxkey = "flux::route";
 
+static struct termios orig_term;
+static bool term_needs_restore = false;
+
+static void save_terminal_state (void)
+{
+    if (isatty (STDIN_FILENO)) {
+        tcgetattr (STDIN_FILENO, &orig_term);
+        term_needs_restore = true;
+    }
+}
+
+static void restore_terminal_state (void)
+{
+    if (term_needs_restore) {
+        /*
+         *  Ignore SIGTTOU so we can write to controlling terminal
+         *   as background process
+         */
+        signal (SIGTTOU, SIG_IGN);
+        tcsetattr (STDIN_FILENO, TCSADRAIN, &orig_term);
+        /* https://en.wikipedia.org/wiki/ANSI_escape_code
+         * Best effort: attempt to ensure cursor is visible:
+         */
+        printf ("\033[?25h\r\n");
+        fflush (stdout);
+        term_needs_restore = false;
+    }
+}
+
+
 static void completion_cb (flux_subprocess_t *p)
 {
     struct proxy_command *ctx = flux_subprocess_aux_get (p, "ctx");
@@ -318,7 +349,9 @@ static int cmd_proxy (optparse_t *p, int ac, char *av[])
 
     /* Start reactor
      */
+    save_terminal_state ();
     if (flux_reactor_run (r, 0) < 0) {
+        restore_terminal_state ();
         log_err ("flux_reactor_run");
         goto done;
     }
```

**Comment by garlick** [comment]
That seems harmless to me.  I guess if there is no controlling tty it silently fails which should be OK?

If it helps, let's get it in!

**Comment by grondo** [comment]
If there is no controlling tty is should do nothing, but I guess I should try that.

**Comment by grondo** [comment]
Hm, this additional approach (mentioned above) also seems to help (in rough draft state)

By default we shut down the usock_server and router and wait for the child to complete, instead of simply exiting and allowing the shell or child process to be inherited by PID 1. This might be nice because it would allow a user to finish possible interactive work in the child shell (after `flux-proxy` itself exits and the shell is reparented, the shell becomes unusable)

A new `-k, --kill` option could be introduced to instead send SIGTERM to the child of `flux proxy` when the
connection is lost. e.g.:

```console
ƒ(s=1,d=0,builddir) grondo@asp:~/git/fc.git$ flux proxy -k ƒMnbjrioSf
ƒ(s=1,d=1,builddir) grondo@asp:~/git/fc.git$ flux --parent job cancel $(flux getattr jobid)
ƒ(s=1,d=1,builddir) grondo@asp:~/git/fc.git$ flux-proxy: Lost connection, terminating
flux-proxy: Terminating child processes...
exit

ƒ(s=1,d=0,builddir) grondo@asp:~/git/fc.git$
```
```console
ƒ(s=1,d=0,builddir) grondo@asp:~/git/fc.git$ flux proxy ƒMnqJumXaF
ƒ(s=1,d=1,builddir) grondo@asp:~/git/fc.git$ flux --parent job cancel $(flux getattr jobid)
ƒ(s=1,d=1,builddir) grondo@asp:~/git/fc.git$ flux-proxy: Lost connection, terminating

lflux: flux_open: No such file or directory
sflux: flux_open: No such file or directory
ƒ(s=,d=,builddir) grondo@asp:~/git/fc.git$ exit
exit

ƒ(s=1,d=0,builddir) grondo@asp:~/git/fc.git$
```

```diff
diff --git a/src/cmd/builtin/proxy.c b/src/cmd/builtin/proxy.c
index ad5cc0c95..ff080f009 100644
--- a/src/cmd/builtin/proxy.c
+++ b/src/cmd/builtin/proxy.c
@@ -351,8 +351,21 @@ static int cmd_proxy (optparse_t *p, int ac, char *av[])
      */
     save_terminal_state ();
     if (flux_reactor_run (r, 0) < 0) {
+        if (errno == ECONNRESET)
+            log_msg ("Lost connection, terminating");
+        else
+            log_err ("flux_reactor_run");
+        if (optparse_hasopt (p, "kill")) {
+            log_msg ("Terminating child processes...");
+            flux_future_destroy (flux_subprocess_kill (ctx.p, SIGTERM));
+        }
+        usock_server_destroy (ctx.server); // destroy before router
+        ctx.server = NULL;
+        router_destroy (ctx.router);
+        ctx.router = NULL;
+        // Wait for child to normally terminate
+        flux_reactor_run (r, 0);
         restore_terminal_state ();
-        log_err ("flux_reactor_run");
         goto done;
     }
 done:
@@ -369,6 +382,8 @@ done:
 static struct optparse_option proxy_opts[] = {
     { .name = "force",  .key = 'f',  .has_arg = 0,
       .usage = "Skip version check when connecting to Flux broker", },
+    { .name = "kill",  .key = 'k',  .has_arg = 0,
+      .usage = "Send SIGTERM to child processes when connection is lost", },
     OPTPARSE_TABLE_END,
 };
 

```

**Comment by grondo** [comment]
Should have been closed by #4058, but issue number was typoed in commit message.
